common/flatpak-dir.c

```diff
@@ -213,6 +215,126 @@ flatpak_get_system_default_base_dir_location (void)
+  groups = g_key_file_get_groups (keyfile, &n_groups);
+  for (i = 0; i < n_groups; i++)
+    {
+      g_autofree char *id = NULL;
+      g_autofree char *path = NULL;
+      size_t len;
+
+      if (!g_str_has_prefix (groups[i], "Installation \""))
+        continue;
+
+      id = g_strdup (&groups[i][14]);
+      if (!g_str_has_suffix (id, "\""))
+        continue;
+
+      len = strlen (id);
+      if (len > 0)
+        id[len - 1] = '\0';
+
+      path = g_key_file_get_string (keyfile, groups[i], "Path", &error);
+      if (error != NULL)
+        g_warning ("Unable to get path for installation '%s': %s\n", id, error->message);
+      else
+        g_ptr_array_add (locations, g_file_new_for_path (path));
+    }
+}
+
+static GPtrArray *
+system_locations_from_configuration (void)
+{
+  g_autoptr(GPtrArray) locations = NULL;
+  g_autoptr(GFile) conf_dir = NULL;
+  g_autoptr(GFileEnumerator) dir_enum = NULL;
+  g_autoptr(GError) error = NULL;
+
+  locations = g_ptr_array_new_with_free_func (g_object_unref);
+
+  if (!g_file_test (SYSCONF_INSTALLATIONS_DIR, G_FILE_TEST_IS_DIR))
+    {
+      g_debug ("No installations directory in %s", SYSCONF_INSTALLATIONS_DIR);
+      goto out;
+    }
+
+  conf_dir = g_file_new_for_path (SYSCONF_INSTALLATIONS_DIR);
+  dir_enum = g_file_enumerate_children (conf_dir,
+                                        G_FILE_ATTRIBUTE_STANDARD_NAME "," G_FILE_ATTRIBUTE_STANDARD_TYPE,
+                                        G_FILE_QUERY_INFO_NONE,
+                                        NULL, &error);
+  if (error != NULL)
+    {
+      g_debug ("Unexpected error retrieving extra installations in %s: %s",
+               SYSCONF_INSTALLATIONS_DIR, error->message);
+      goto out;
+    }
+
+  while (TRUE)
+    {
+      GFileInfo *file_info;
+      GFile *path;
+      const char *name;
+      guint32 type;
+
+      if (!g_file_enumerator_iterate (dir_enum, &file_info, &path,
+                                      NULL, &error))
+        {
+          g_debug ("Unexpected error reading file in %s: %s",
+                   SYSCONF_INSTALLATIONS_DIR, error->message);
+          continue;
+        }
+
+      if (file_info == NULL)
+        break;
+
+      name = g_file_info_get_attribute_byte_string (file_info, "standard::name");
+      type = g_file_info_get_attribute_uint32 (file_info, "standard::type");
+
+      if (type == G_FILE_TYPE_REGULAR && g_str_has_suffix (name, ".install"))
+        {
+          g_autofree char *path_str = g_file_get_path (path);
+          append_locations_from_config_file (locations, path_str);
+        }
+    }
+
+ out:
+  return g_steal_pointer (&locations);
+}
+
+GPtrArray *
+flatpak_get_system_base_dir_locations (void)
+{
+  static gsize array = 0;
+
+  if (g_once_init_enter (&array))
+    {
+      gsize setup_value = 0;
+      setup_value = (gsize) system_locations_from_configuration ();
+      g_once_init_leave (&array, setup_value);
+     }
```

Comment:
GNOME Software will most once to get the list of installations, or maybe N+1 (where N is the number of installations) in case it needs to fetch every installation by ID, independently, so it's probably a good idea not to cache it for that reason, at least for now.

At the same time, I doubt adding/removing installations will be a common enough operation 
that we have to worry about needing to restart GS to pick them up, and there are quite a few places where flatpak_dir_get_system_list() gets called in flatpak-utils.c and some of the CLI commands (e.g. info, run), so I'm still feeling like we should live the cache in place.

What do you think?